src/core/crypto.c

```diff
@@ -1719,7 +1719,53 @@ QuicCryptoCustomCertValidationComplete(
             "Custom cert validation failed.");
         QuicConnTransportError(
             QuicCryptoGetConnection(Crypto),
-            QUIC_ERROR_CRYPTO_ERROR(0xFF & CXPLAT_TLS_ALERT_CODE_BAD_CERTIFICATE));
+            QUIC_ERROR_CRYPTO_ERROR(0xFF & CXPLAT_TLS_ALERT_CODE_BAD_CERTIFICATE)); //
+    }
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+QuicCryptoCustomTicketValidationComplete(
+    _In_ QUIC_CRYPTO* Crypto,
+    _In_ BOOLEAN Result
+    )
+{
+    if (!Crypto->TicketValidationPending) {
+        return;
+    }
+
+    Crypto->TicketValidationPending = FALSE;
+    if (Result) {
+        Crypto->TlsState.ReadKey = QUIC_PACKET_KEY_INITIAL;
+        Crypto->TlsState.WriteKey = QUIC_PACKET_KEY_INITIAL;
+        
+        Crypto->TlsState.BufferLength = 0;
+        Crypto->TlsState.BufferTotalLength = 0;
+        Crypto->TlsState.BufferOffsetHandshake = 0;
+        Crypto->TlsState.BufferOffset1Rtt = 0;
+        Crypto->TlsState.ReadKeys[1] = NULL;
+        Crypto->TlsState.ReadKeys[2] = NULL;
+        Crypto->TlsState.ReadKeys[3] = NULL;
+        Crypto->TlsState.ReadKeys[4] = NULL;
+        Crypto->TlsState.ReadKeys[5] = NULL;
+
+        Crypto->TlsState.WriteKeys[1] = NULL;
+        Crypto->TlsState.WriteKeys[2] = NULL;
+        Crypto->TlsState.WriteKeys[3] = NULL;
+        Crypto->TlsState.WriteKeys[4] = NULL;
+        Crypto->TlsState.WriteKeys[5] = NULL;
+        Crypto->ResultFlags = 0;
+
+        QUIC_CONNECTION* Connection = QuicCryptoGetConnection(Crypto);
+        QUIC_CONFIGURATION* Configuration = Connection->Configuration;
+        Crypto->Initialized = FALSE;
+        uint8_t* Buffer = Crypto->RecvBuffer.Buffer;
+        QuicCryptoInitialize(Crypto);
+        Crypto->RecvBuffer.WrittenRanges.UsedLength = 1; // To pass QuicRecvBufferRead
+        Crypto->RecvBuffer.Buffer = Buffer; // To pass "if" next to QuicCryptoTlsGetCompleteTlsMessageLength
+        QuicCryptoInitializeTls(Crypto, Configuration->SecurityConfig, Connection->HandshakeTP);
```

Comment:
See my comment above. In the success case, you should just clear the pending flag and then call `QuicCryptoProcessDataComplete`. All this clearing of keys and re-initialization is unnecessary in the success path. You will have to do something along these lines for the failure path though.

Reply:
oh, maybe works😮